proposals/002-text-utf-default.md

```diff
@@ -0,0 +1,156 @@
+This proposal is for the migration of the `text` package from its
+default UTF-16 encoding to UTF-8. The lack of UTF-8 as a default in the
+`text` package is a pain point raised by the Haskell Community and
+industry partners for many years. The Haskell Foundation is uniquely
+positioned to effect a change like this, granted there is an appetite
+for breakage, and that such a project is well-socialized and planned.
+During the meetings of the Haskell Foundation Tech Agenda Track, we
+identified the pros and cons of such a migration against historical
+attempts, and agreed that it was within our power to do.
+
+Further, we solicited both community and industry feedback to gauge the
+appetite for breakage, and what stakeholders would be affected by the
+changes. Andrew Lelechenko has offered to lead the implementation of
+this project.
+
+# Motivation
+
+-   UTF-16 by default requires that all Text values pay a premium for serialization. Arguably, the performance impact of Text is flipped
+    upside-down: most text is UTF-8, and Haskell devs pay an undue cost when working with the wrong default.
+
+-   UTF-8 is the industry standard and by far the most common text encoding, with roughly 97% of web pages existing in UTF-8. The
+    existing UTF-16 default imposes an additional hurdle to working with the vast majority of web content on earth.
+
+-   Many systems in Haskell are UTF-8 by default (e.g. Haddock)
+
+# Goals
+
+-   Solicit feedback from community members and industry to gauge appetite for such a change.
+
+-   Provide an implementation, migration, and delivery plan for changing the default encoding of `text` from UTF-16 to UTF-8.
+
+-   Ensure stakeholders (e.g. GHC, Cabal, Stack, boot libs) have ample time to migrate and address any bugs.
+
+-   Implementation should not significantly alter the performance characteristics of the base `text` library within some tolerance
+    threshold.
+
```

Comment:
Are there any performance improvement goals? For example: a 10% speedup of reading HTML delivered off the web (presumably in UTF-8 format).

Reply:
There is potential for performance boost, as most on the wire text is UTF8
nowadays, so it only needs to be validated not necessarily copied.

However, if data is read as `ByteString` (backed by `ForeignPtr`) (as it often is)
then we still would need to copy to `ByteArray#` backing up `Text`.

So my gut feeling is that performance gains will arise mostly from reduced memory pressure.
That said, for e.g. Russian or Greek text the performance won't be any better. Or any other script which is mostly outside of ASCII (note, not even 256 first codepoints) range.

E.g. in finnish which is mostly ASCII but has common öä:

```
-rw-rw-r--  1 phadej phadej   597661 May 13 22:45 kalevala.txt
-rw-rw-r--  1 phadej phadej  1150892 May 13 22:46 kalevala16.txt
```

the UTF-8 version is 51% of UTF-16 text.

However with Russian (this is a word list file)

```
-rw-rw-r--  1 phadej phadej 36530302 May 13 22:50 russian16.txt
-rw-rw-r--  1 phadej phadej 34993984 May 13 22:50 russian8.txt
```

the UTF-8 version is 95% of UTF-16 (the saving comes from line feeds and space characters). Almost the same.

---

I.e. any performance changes are not given.
E.g. Z-Data performance gains are (as far as I can tell) from copy avoidance, but `ForeignPtr` vs. `ByteArray#` ruin that chance.

Reply:
> However, if data is read as `ByteString` (backed by `ForeignPtr`) (as it often is)
then we still would need to copy to `ByteArray#` backing up `Text`.

Why must `Text` be backed by `ByteArray#`? What are the cons of having a `Text` where decoding from UTF-8 makes no copy?

Reply:
Thanks @phadej for this explanation. I'm not surprised here -- performance is always variable. But there still can be performance goals. For example: "Reduction of 50% of memory usage on ASCII text". We're restricting the input here -- ASCII text only -- but still stating an admirable goal. I bet we can do even better, though, by stating performance goals over averages (while limiting regressions of any particular test).

The underlying reason we want this proposal is to improve performance of `text` (I think), so it seems short-sighted not to state a goal of improving performance of `text`, suitably qualified to account for reality.

Reply:
@Lysxia, one can argue in different directions. Why `ByteString` have to be backed by `ForeignPtr`. The only good answer I know is to support `mmap`, as we cannot create `ByteArray#` from external memory (why can't we?).

OTOH, it's extremely tempting to be able to `mmap` textual file into (strict) `Text`. That **will** save time.

Reply:
@goldfirere the ASCII text is very West centric view. Note, we talk about `Text`, which is used for processing text as in books, chat messages (!!!) etc.

"programming languages" may be important, but in "real world" that is still a niche use case.

Reply:
Recall the representation:
- header 8 bytes
- offset + length: 2x8bytes = 8bytes
- payload: at least 2 bytes per codepoint

Saving 50% memory usage is technically impossible. To get close you need to have big strings. (Like whole comments e.g.) and to benefit from slicing you need overlapping slices.

Slices are useful so `drop 1` etc. are fast - they don't save memory.

---

What I try to say, without success, is that if you are concerned with performance then there cannot be a single `Text` representation which is just best. 

A simple example is that people don't use `Data.Text.Builder`, but just concatenate strict  `Text` values (in logging). That is just bad, O(n^2) copying. Where simple `"foo" ++ show aNumber ++ "bar" :: String` is actually not that bad!

Reply:
I agree that ASCII text is very limiting. So perhaps that's the wrong metric. But we need to identify *some* goal here, otherwise we won't know whether we've been successful.

As it stands, the proposal's net effect is "this proposal will change performance of `text`". I don't doubt that performance will change. But if that's the best net effect we can muster, it doesn't seem, to me, that this proposal is worth investing time and energy behind. On the other hand, if we can state a goal of improving (average) performance across a range of inputs, that sounds exciting. Maybe we'll meet the goal, and maybe we won't, but having a goal at least frames the discussion so we can evaluate, later, whether merging is a good idea.

Maybe more simply: the fact that performance is variable across inputs is not (to me) a reason not to measure performance.

Reply:
> What I try to say, without success, is that if you are concerned with performance then there cannot be a single Text representation which is just best.

I think this has been successful. I appreciate this, and I appreciate that, whatever we do, some applications of `Text` will get faster and some will get slower. That's all good. But I do hope that we can expect an improvement **on average** (for a suitably fair definition of average).

Reply:
In addition to stakeholders agreeing on a set of benchmarks worthy of both _optimizing against_ and _testing against for regressions_, I think @phadej 's comments make it clear that micro-benchmarks won't be sufficient... you'll need to measure the entire life-cycle: serialization, deserialization, additional copying, effects on memory and runtime of sharing/slicing, FFI, etc.

Reply:
I think the argument is that there will be variable impact across a wide range of different applications, and therefore picking _beforehand_ a concrete metric is not the right approach.

If one is doing the "textbook scientific method" (that no scientists actually use) where you fix a hypothesis and then perform an experiment, then this is what you need to do. But that's not the way engineering works, and its not arguably even how lots of science works. When you do a drug trial, you statistically validate against the null hypothesis, but you don't fix a "amount of cancer cured" threshold before running the trial.

So I would say there is strong suspicion this will be helpful in certain large applications (but perhaps not others) and on measuring the impact after the fact, a decision can be made on what to do as the default.

Arguably, the fact that there isn't a single quantifiable metric to reduce everything to, but instead this range of tradeoffs and unknowns is partly what has stalled similar projects for so long. I'd say that there's sufficiently broad sentiment to try to do this that its worth it to first do it, then just see what happens, and then proceed based on that, and also now there will be a "definitive answer" as to why things are why they are, based on a significant investment of resources both in development and in benchmarking.

Reply:
@gbaz, I agree. But that should be written explicitly into the proposal in some way.

Otherwise when things are done, sunk cost etc and the change is pushed forward no matter what are result.

Experiments should be allowed to fail. Currently this is not setup as such experiment.

Reply:
All I'm arguing for here is a clear metric by which we will evaluate whether this effort has been successful. Perhaps I have over-focused on percentages -- maybe it would instead be better to focus on the choice of benchmarks. In any case, I'm in support of looking at the results holistically later to decide whether this project has been successful. However, we need to decide the metric of evaluation (that is, the benchmarks) ahead of time. Otherwise, as @phadej says, we will be able to declare success no matter what happens.

Furthermore, I still think having goal metrics (not just chosen benchmarks) is a good idea. Despite the various claims here that it is impossible to get such goals, I've seen two plausible ones floated: @davean says (https://github.com/haskellfoundation/tech-proposals/pull/1#discussion_r632036534) that memory usage of UTF-8 should be roughly half that of UTF-16 for ASCII text. And @chrisdone describes (https://github.com/haskellfoundation/tech-proposals/pull/1#issuecomment-840870519) that the conversion takes 1us/kb, even providing a concrete test (thank you!). These observations can provide the basis of goals.

Here's another way to think about it: We, as a community, are choosing whether to invest in the idea presented here. At what point will our investment be worth it? If 1% is shaved off the running time of one application, but all others suffer a 10% regression, is this worth pursuing? (I imagine not.) If the geometric mean of a suite of applications shows no improvement, some applications improve by 2%, and a few regress by 20%, is it worth it?  (This is a harder one.) This thinking gives us a different way to phrase the goal: identify the minimum improvement we should expect to see in order to declare success. Hopefully, we'll far surpass it. Maybe we'll fall short -- but there might be mitigating reasons to accept the change anyway.

To connect most directly with @gbaz's argument: the development of the covid vaccine is about as experimental as you can get. Yet it still had a goal: in the US, the goal of 50% efficacy was frequently mentioned (where efficacy was well-defined in terms of double-blind trials and particular ways of measuring). I doubt anyone had any idea of whether that goal could be achieved -- but it was identified as a goal so that we could say, if a vaccine arrived with only 40% efficacy, that we should keep searching and not do a major campaign to get everyone vaccinated. We have, amazingly, far surpassed that goal with several highly effective vaccines. Just as I hope that the work proposed here will far surpass a modest goal we might set.

Reply:
" If 1% is shaved off the running time of one application, but all others suffer a 10% regression, is this worth pursuing? (I imagine not.)" Ok, but how will we know this until the library is written and we then test it against a bunch of stuff?

I don't think setting specific percentages at all helps here. One could state general expectations -- that it improves some things by a noticeable amount (likely memory-dependent things and ffi dependent things), and generally doesn't cause regressions elsewhere. The only real metric that I think matters is "the good outweighs the bad in a way that on the whole improves things for most users." 

Expectations of performance increase in any more detail are more _motivations_ than criteria for success.

Reply:
> Ok, but how will we know this until the library is written and we then test it against a bunch of stuff?

We won't know this. That's the whole point: we state what we'd like to achieve and then try to achieve it. If we can't, that doesn't mean failure -- it just means that we have to assess whether what we have achieved is worthwhile. So, maybe we're actually arguing for mostly the same thing: I might be able to interpret your viewpoint as "set the goal to be 100% improvement of all performance metrics". Of course, that goal is impossible, and so whatever we achieve would have to be debated later. My stance is to set an achievable goal, so that we can short-circuit the later discussion. Note that discussing this later is fraught: as the work will already have been done, it is much harder to say "we shouldn't do it".

One thing in common to both approaches, though, is that we can at least set out the benchmarks now. As I said before, I think percentages aren't the most important point -- choosing benchmarks is.

>  some things

What set are these things drawn from? This gets to my desire for benchmarks.

> noticeable amount

What is noticeable? My guess is that you would agree that a 0.001% change is not "noticeable", while a 5% change is. If that's the case, then you have an implicit percentage expectation. I am trying to make that implicit percentage explicit.

> generally doesn't cause regressions

What's a regression? Is a slip of 0.001% a regression? Is a slip of 5% a regression? Again, you probably have an implicit percentage expectation. 

What's "generally"? I interpret this to mean that some regressions somewhere are OK, as long as the regressions aren't everywhere. Maybe this means that you expect the geometric mean performance measurements of xyz to be 0 or better? (And what's "xyz"?)

> Expectations of performance increase in any more detail are more motivations than criteria for success.

I agree here, and thanks for pointing this out. This whole thread started by my asking about the motivation for the proposal. (That's not clear from the history on this ticket, but the comment at the root of this thread was actually internally spun off from https://github.com/haskellfoundation/tech-proposals/pull/1#discussion_r631985001.) While I still think stating a percentage improvement on certain measurements as a goal is fruitful, I'd be content with stating it as a motivation.

I guess, zooming out, that what I'm really after is in making certain implicit assumptions more explicit, lest we all get surprised later when our own internal implicit assumptions differ from others'. The assumptions I'm referring to are things like what measurements to take, what benchmarks to run, what "noticeable" means, what "regression" means, etc. We all have some intuitive sense of this, but let's write it down -- because my guess is that each of us will have slightly different intuitions, and a diversity of unstated assumptions now will cause unhappiness later.

Reply:
> We all have some intuitive sense of this, but let's write it down -- because my guess is that each of us will have slightly different intuitions, and a diversity of unstated assumptions now will cause unhappiness later.

Let me suggest the exact opposite. Suppose there is an overall 20% improvement with no regressions over 1%. Nobody would say "let's not take advantage of that". So the differences in intuitions and assumptions would not matter!

Now if we spent a lot of time now making sure we hashed out all those different intuitions and assumptions, would that change things? No! It would just be a lot of time now arguing about imaginary percentages to get on the same page, and it would turn out those disputes did not matter, because the facts presented a clear picture.

You're right that there are going to be a ton of places where there's room for potential debate after we have some numbers and results in. My view is I think it is _extremely difficult_ to gauge which of those places will be the place debate actually occurs _until_ those results are in. So therefore I would rather not have debates and speculation about things that will not matter, and have the discussions concretely, later, when we know that they are about actual results and not speculative results. I suspect this approach will save a lot of brain cells.

Reply:
I've updated sections on Motivation and Performance impact.

Reply:
> So my gut feeling is that performance gains will arise mostly from reduced memory pressure.
> That said, for e.g. Russian or Greek text the performance won't be any better. Or any other script which is mostly outside of ASCII (note, not even 256 first codepoints) range.

In the spirit of full disclosure, the scripts used by the majority of languages originating from east of the Mediterranean and now spoken by almost half of the human race require 3 bytes in UTF-8 vs 2 in UTF-16, so the memory pressure will be worse.

_However_, as UTF-8 Everywhere [points out](http://utf8everywhere.org/#asian), real-world documents typically don't increase much in size with UTF-8 vs UTF-16 because they include a lot of ASCII text, in newlines, spaces, punctuation, markup, quotations and so on. This wouldn't apply to shorter pieces of text such as keys, names, words and phrases, but an application would always have the option of using a UTF-16-based text library for that. (A case could even be made for using UTF-16 for `ShortText` since those will have to be created by copying rather than slicing, since the storage model is different, so transcoding won't matter.)

Reply:
Good points @gbaz, and your argument moved the needle for me. In any case, @Bodigrim's update delivered on what I was hoping for: describing a clear win (> 50% !) for some functions and no regressions elsewhere (defined on a specified testsuite).

Shall we Resolve this conversation then? Given how many have contributed so thoughtfully, I didn't want to just close it out unilaterally. But perhaps the next person to come along, if you agree, just go ahead and close it.

Reply:
> In the spirit of full disclosure, the scripts used by the majority of languages originating from east of the Mediterranean and now spoken by almost half of the human race require 3 bytes in UTF-8 vs 2 in UTF-16, so the memory pressure could be worse in some cases.

Thanks for this information -- I don't know much about character encodings and falsely assumed that the 3-byte UTF-8 slots were for more exotic characters than those used in everyday writing. I apologize.

Reply:
> Why must `Text` be backed by `ByteArray#`? What are the cons of having a `Text` where decoding from UTF-8 makes no copy?

The main drawback is that pinned memory causes heap fragmentation: a single `ByteString` can easily keep 1Mb block allocated, because GC is not allowed to move it around. 

Indeed it would be nice to back `ByteString` and `Text` by the same structure: either `ByteArray` or `ForeignPtr`, but one way or another this is a far bigger change than UTF-8 transition. It seems that current consensus is that it's up to `ByteString` to change, but anyways it is out of scope of this proposal.

Reply:
Aside: whenever I see the fragmentation issue brought up I wonder whether anyone has explored ByteString allocation with jemalloc rather than libc. Its main selling point is low fragmentation. Rust uses jemalloc.

Reply:
@chrisdone `ByteString`s are "malloced" from GHC RTS heap.

Reply:
Granted, but they consist of a foreign pointer any way; therefore they can take any allocation strategy, and in many cases do when strings come from C via the `unsafePack*` functions.